torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -483,6 +487,9 @@ def __init__(
       gc.collect()
       xm.mark_step()
 
+    if save_shard_metadata:
+      self._register_state_dict_hook(self.get_shard_metadata)
```

Comment:
If I understand correctly, this `_register_state_dict_hook` will cause `self.get_shard_metadata` to be called on **every** FSDP module (not just the root FSDP module)? I believe we would only want to call `self.get_shard_metadata` on the root instance

The implementation in this PR seems to lead to some problems in nested FSDP as well as multiple root FSDP instances. Consider the following model hierarchy
```
class A(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear_fsdp1 = FSDP(torch.nn.Linear(10, 3))
        self.linear_regular = torch.nn.Linear(10, 4)

class B(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.a = FSDP(A())
        self.linear_fsdp2 = FSDP(torch.nn.Linear(10, 5))

b = B()
```
We will only want to `self.get_shard_metadata` on `b.a` and `b.linear_fsdp2` since they are the root FSDP instances (but not `b.a.linear_fsdp1` since it's not a root FSDP). But when calling `b.state_dict()`, it cannot guarantee which FSDP's metadata will be saved to `destination["shard_metadata"]`.